Add Encode(Span<char>) API

steveharter opened the pull request:
Fixes https://github.com/dotnet/corefx/issues/39523

For performance adds the API linked above for encoding char. Also changes the semantics of System.Text.Json to use the replacement character for Utf16 for bad surrogate pairs instead of throwing - this makes it consistent with Utf8 and the default behavior of System.Text.Encoding.Json.

Improves performance of Utf16 encoding (with custom encode) by ~25% by avoiding Utf16->Utf8->Utf16 conversions:
```
Before
|            Method |       Mean |    Error  |   StdDev  |     Median |        Min |        Max | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
| EscapeUtf16Custom | 1,745.3 ns | 19.710 ns | 16.459 ns | 1,749.1 ns | 1,719.2 ns | 1,774.9 ns |      0.2635 |           - |           - |             1.66 KB |
|  EscapeUtf8Custom | 1,606.9 ns |  8.175 ns |  6.827 ns | 1,607.7 ns | 1,593.5 ns | 1,618.8 ns |      0.2657 |           - |           - |             1.66 KB |

After
| EscapeUtf16Custom | 1,285.4 ns | 17.460 ns | 16.332 ns | 1,279.2 ns | 1,264.2 ns | 1,312.6 ns |      0.2684 |           - |           - |             1.66 KB |
|  EscapeUtf8Custom | 1,150.1 ns |  9.990 ns |  9.344 ns | 1,149.0 ns | 1,133.8 ns | 1,167.3 ns |      0.2702 |           - |           - |             1.66 KB |
```

Also by changing code in System.Text.Json, improves performance of Utf8 escaping when no escaping needs to occur (~15% faster), or occurs later in the string:
```
Before
|            Method | Formatted | SkipValidation |     Escaped |      Mean |     Error |    StdDev |    Median |       Min |       Max | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|------------------ |---------- |--------------- |------------ |----------:|----------:|----------:|----------:|----------:|----------:|------------:|------------:|------------:|--------------------:|
|  WriteStringsUtf8 |     False |          False |  AllEscaped | 54.647 ms | 0.1615 ms | 0.1432 ms | 54.612 ms | 54.468 ms | 54.922 ms |           - |           - |           - |               120 B |
|  WriteStringsUtf8 |     False |          False |  OneEscaped |  9.646 ms | 0.0793 ms | 0.0742 ms |  9.615 ms |  9.583 ms |  9.812 ms |           - |           - |           - |               120 B |
|  WriteStringsUtf8 |     False |          False | NoneEscaped |  6.866 ms | 0.2275 ms | 0.2620 ms |  6.759 ms |  6.555 ms |  7.303 ms |           - |           - |           - |               120 B |


After
|  WriteStringsUtf8 |     False |          False |  AllEscaped | 52.957 ms | 0.5980 ms | 0.5594 ms | 52.929 ms | 52.145 ms | 53.978 ms |           - |           - |           - |               120 B |
|  WriteStringsUtf8 |     False |          False |  OneEscaped |  8.678 ms | 0.0992 ms | 0.0879 ms |  8.688 ms |  8.529 ms |  8.845 ms |           - |           - |           - |               120 B |
|  WriteStringsUtf8 |     False |          False | NoneEscaped |  5.604 ms | 0.0700 ms | 0.0655 ms |  5.610 ms |  5.495 ms |  5.757 ms |           - |           - |           - |               120 B |
```

FWIW the ASCII escaping in System.Text.Json is ~25% faster than System.Text.Encoding.Web.JavaScriptEncoder - thus the reason why there is some duplicate functionality between the two; S.T.J will only call S.T.E.W for non-ASCII or when using a non-default encoder.

GrabYourPitchforks commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs:
```diff
@@ -132,9 +139,18 @@ public virtual string Encode(string value)
         // and the underlying Mono bug: https://bugzilla.xamarin.com/show_bug.cgi?id=36052.
         // If changing the signature of this method, ensure this issue isn't regressing on Mono.
-        private unsafe int EncodeIntoBuffer(char* buffer, int bufferLength, char* value, int valueLength, int firstCharacterToEncode)
+        private unsafe OperationStatus EncodeIntoBuffer(
```
Is the comment above this method still relevant? Seems like the issue has been fixed for a few years. If this comment isn't still relevant, please delete it.

steveharter replied:
Sure. The issue was fixed in Mono 3.5 years ago, so deleting the comment.

However, I don't plan on changing the argument ordering since it was (source, destination):
`char* value, int valueLength, int firstCharacterToEncode, char* buffer, int bufferLength` and the current API is (destination, source) which is consistent with other similar APIs.

GrabYourPitchforks commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs:
```diff
@@ -132,9 +139,18 @@ public virtual string Encode(string value)
         {
-            int totalWritten = 0;
+            char* originalBuffer = buffer;
+            charsWritten = 0;
```
Nit: You'll get a slight perf boost if you create a temporary local for this value and don't assign the _out_ parameter until the very end of the method. This could allow the local to be eligible for enregistration.

steveharter replied:
Sure. Thanks

steveharter replied:
Update: I will keep the code as-is. On my local build (Windows + release) it is faster to write to the `out` parameter. I verified with a benchmark.

The three cases with a local:
```cs
charsWrittenLocal += firstCharacterToEncode;
00007FF835CF6299  mov         esi,r15d  
---
charsWrittenLocal++;
00007FF83BAE4A97  mov         r9d,dword ptr [rsp+6Ch]  
00007FF83BAE4A9C  inc         r9d  
00007FF83BAE4A9F  mov         dword ptr [rsp+6Ch],r9d  
---
                    charsWrittenLocal += charsWrittenThisTime;
00007FF83BAE4B39  mov         r10d,dword ptr [rsp+6Ch]  
00007FF83BAE4B3E  add         r10d,dword ptr [rsp+50h]  
00007FF83BAE4B43  mov         dword ptr [rsp+6Ch],r10d  
```

With the `out` parameter:
```cs
                charsWritten += firstCharacterToEncode;
00007FF835CF47FC  add         dword ptr [r14],r12d  
---
 charsWritten++;
00007FF835AD7AD4  inc         dword ptr [r14]  
---
                    charsWritten += charsWrittenThisTime;
00007FF835AD7B6E  mov         eax,dword ptr [rsp+48h]  
00007FF835AD7B72  add         dword ptr [r14],eax  
```

GrabYourPitchforks commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs:
```diff
@@ -177,41 +193,51 @@ private unsafe int EncodeIntoBuffer(char* buffer, int bufferLength, char* value,
-                int nextScalar = UnicodeHelpers.GetScalarValueFromUtf16(firstChar, null, out wasSurrogatePair);
-                if (!TryEncodeUnicodeScalar(nextScalar, buffer, bufferLength, out charsWritten))
+                int nextScalar = UnicodeHelpers.GetScalarValueFromUtf16(firstChar, null, out wasSurrogatePair, out bool wasUnmatchedSurrogate);
+                if (!isFinalBlock && wasUnmatchedSurrogate)
```
I believe this check is incorrect. `NeedMoreData` should only be returned if `isFinalBlock` is _false_ and the very last character of the buffer is a UTF-16 high surrogate character.

steveharter replied:
I believe the code is doing what you state. `!isFinalBlock` + `wasUnmatchedSurrogate`.

GrabYourPitchforks replied:
Won't _wasUnmatchedSurrogate_ also be true if the very last character of the buffer is a UTF-16 low surrogate character?

steveharter replied:
Yes thanks. Changed the variable name to make it more clear.

GrabYourPitchforks replied:
Just to clarify - I don't have a strong preference on what the variable name is called. :)

All I wanted to check was that if _isFinalBlock_ is _false_ and the very last char is a high surrogate character, the API returns `NeedsMoreData`. Any other combination (_isFinalBlock_ is _true, or the very last char is a low surrogate character) should not return `NeedsMoreData`. I didn't see a unit test for this condition so I wanted to call it out as something to double-check.

steveharter replied:
I renamed `wasUnmatchedSurrogate` to `needMoreData` and changed the logic so `needMoreData` returns true\false as you state.

GrabYourPitchforks commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs:
```diff
@@ -719,12 +779,18 @@ private byte[] GetAsciiEncoding(byte value)
             return encoding;
         }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
```
@AndyAyersMS - I recall you saying something in the past re: methods that throw exceptions shouldn't be marked _noinlining_, since this attribute could prevent the JIT from recognizing the call site as invoking a no-return method. Is this still the recommendation?

AndyAyersMS replied:
Yes, it's still the case.

steveharter replied:
I will remove the NoInlining here.

However when using the "throw helper" pattern which helps reduce the [size of code for generics methods](https://github.com/dotnet/coreclr/issues/22400) there are several cases (at least in System.Text.Json) where NoInlining is specified. I assume because otherwise it is possible the throwhelper method would just be inlined (?).

AndyAyersMS replied:
Without the NoInlining, throw helpers will not get inlined.

The jit knows how to recognize these methods and will avoid inlining them and also mark the throw helper call as cold so that the call site is moved to the end of the method. This latter bit gives better hot code locality and usually eliminates a taken branch.

In principle you can use the jit-diff tool from [jitutils](https://github.com/dotnet/jitutils) to look at the impact on the jit-generated code with and without the attribute(s). This currently requires a checked jit so isn't as CoreFx friendly as one would like. I would expect to see similar size overall with and without the NoInlining attribute on throw helpers, but the code flow in calling methods will be less optimal with the attribute in place.

gfoidl commented on src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs:
```diff
@@ -212,304 +216,28 @@ private static void EscapeNextBytes(byte value, Span<byte> destination, ref int
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Is this really needed? I assume the JIT will inline this method without the attribute too.
Also on L222 (method `IsAsciiValue(char)` the attribute could be removed.

steveharter replied:
I applied the attribute because it was present in the other `IsAsciiValue` method.

I verified locally (Windows release build) that the IsAsciiValue methods are inlined without the attribute.

However there are many (60+) uses of AggressiveInlining in System.Text.Json similar to this -- some with a few lines of code (so not as obvious) but also some one-liners, so I'm reluctant to change without some sort of a plan or guidelines --  @ahsonkhan thoughts on when to include AggressiveInlining?

ahsonkhan replied:
Generally speaking, AggressiveInlining should be used sparingly and only if there is evidence it is necessary and is improving perf. I may have used it too liberally in the past, and if a method is getting inlined without the attribute, we shouldn't add one, especially for new code/changes.

**So, I wouldn't add one here.**

We can consider removing some of the other ones that might have been unnecessary, separately.

AndyAyersMS replied:
It occurs to me that with a slightly nonstandard jit, we can flag cases where `AggressiveInlining` is redundant. In conjunction with PMI (from jitutils) we should be able to spot the majority of these cases in CoreFx and CoreCLR.

The jit currently will bypass profitability analysis when it sees `AggressiveInlining`. So the first nonstandard bit would be to always run profitability analysis and then flag cases that are both profitable and marked. The second would be to surface the results somehow... we can be low-tech here I suppose, and just write to stdout or stderr.

Profitability is contextual (depends on caller and call site) so we'd have to look at what happens at all the sites that invoke a method; if most or all of these would inline without the attribute, then we can arguably remove it.

steveharter replied:
I will remove `AggressiveInlining` for the methods in this file except for this one that still requires it (not inlined by default):
`private static bool NeedsEscaping(char value) => value > LastAsciiCharacter || AllowList[value] == 0;`

gfoidl commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/JavaScriptEncoder.cs:
```diff
@@ -227,5 +228,10 @@ private static unsafe bool TryWriteEncodedSingleCharacter(int unicodeScalar, cha
             return true;
         }
+
+        private static void ThrowHelper_ArgumentNullException(string paramName)
```
This method is used only once, so you could move it to a local function and eliminate the `paramName` (i.e. hard-code `"text"`), so the code gets even smaller, as no string-argument must be created.

steveharter replied:
Yes thanks.

watfordgnf commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs:
```diff
@@ -369,7 +400,7 @@ public unsafe virtual OperationStatus EncodeUtf8(ReadOnlySpan<byte> utf8Source,
                         byte[] encodedBytes = GetAsciiEncoding((byte)nextScalarValue);
 
-                        if (ReferenceEquals(encodedBytes, s_noEscape))
+                        if (ReferenceEquals(encodedBytes, Array.Empty<byte>()))
```
Is `Array.Empty<T>()` okay for reference equality? I noticed https://github.com/dotnet/corefx/pull/39708/files added the suppression of the CA1825 warning to TextEncoder for an empty array allocation.

steveharter replied:
I verified it worked locally. Originally I wasn't sure so I created a singleton. Now I'm having second thoughts :)

I'll revert this change back to what I had previously and add additional comments.

gfoidl commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/JavaScriptEncoder.cs:
```diff
@@ -119,6 +119,7 @@ public unsafe override int FindFirstCharacterToEncode(char* text, int textLength
             {
                 throw new ArgumentNullException(nameof(text));
```
@steveharter I meant it this way:
```c#
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public unsafe override int FindFirstCharacterToEncode(char* text, int textLength)
        {
            if (text == null)
            {
                ThrowHelper_ArgumentNullException();
            }

            return _allowedCharacters.FindFirstCharacterToEncode(text, textLength);

            void ThrowHelper_ArgumentNullException()
            {
                throw new ArgumentNullException(nameof(text));
            }
        }
```

steveharter replied:
I just reverted back to the original; I don't think having the ThrowHelper helps here with perf.

steveharter commented:
Note: the second commit addressing feedback was just rebased onto the original (due to local issues with out-of-sync history causing unnecessary commits being shown)

ahsonkhan commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs:
```diff
@@ -524,6 +557,34 @@ internal static OperationStatus EncodeUtf8Shim(TextEncoder encoder, ReadOnlySpan
+        /// <see langword="false"/> if there is no further source data that needs to be encoded.</param>
+        /// <returns>An <see cref="OperationStatus"/> describing the result of the encoding operation.</returns>
+        /// <remarks>The buffers <paramref name="source"/> and <paramref name="destination"/> must not overlap.</remarks>
+        public virtual OperationStatus Encode(
```
Will this method ever throw? Doesn't look like it, but wanted to verify.

steveharter replied:
It does not throw.

ahsonkhan replied:
Destination is a `Span<char>`. How can it be null?

steveharter replied:
You can assign null to a span -- it just gets treated as empty. And then when the pointer is obtained, an [actual null pointer is returned](https://github.com/dotnet/corefx/blob/e30ab6c16dbd37098dc3083712fe75bf5c8c86cc/src/Common/src/CoreLib/System/ReadOnlySpan.Fast.cs#L157). Then the other code (not modified here) checks for null and throws ArgumentNullException.

However, I think we should just return OperationStatus.DestinationTooSmall instead of throwing.

ahsonkhan replied:
Oh... It's going through the implicit cast from array to span. Generally, structs can't be assigned null  (only default), but the hidden cast here is why it works.
```C#
Span<char> span = null; // Equivalent to Span<char> span = (char[])null;
```

ahsonkhan commented on src/System.Text.Encodings.Web/tests/JavaScriptStringEncoderTests.cs:
```diff
@@ -18,11 +19,29 @@ public partial class JavaScriptStringEncoderTests
             }
+
+            Span<char> destination = new char[12];
+            OperationStatus status = System.Text.Encodings.Web.JavaScriptEncoder.Default.Encode("\U0001f4a9".AsSpan(), destination, out int charsConsumed, out int charsWritten, isFinalBlock: true);
```
At some point, we should probably fix up the namespace of these tests (rename `Microsoft.Framework.WebEncoders` to System.Text.Encodings.Web.Tests`) (outside of this PR, ofc).

ahsonkhan replied:
https://github.com/dotnet/corefx/issues/40073

ahsonkhan commented on src/System.Text.Encodings.Web/tests/JavaScriptStringEncoderTests.cs:
```diff
@@ -239,6 +258,83 @@ public void JavaScriptStringEncode_BadSurrogates_ReturnsUnicodeReplacementChar()
+        }
+
+        [Fact]
+        public void JavaScriptEncoder_BadSurrogates_NeedsMoreData()
```
What happens if the input/output spans overlap? Might be worth adding a test for that.

steveharter replied:
I believe it is undefined; the API doc states this must not happen. Do we guard or test against this in any other case?

ahsonkhan replied:
We recently updated `MemoryExtensions.Trim`, to explicitly guard against that and throw. Something to consider - however I don't think it is absolutely necessary. The trim case was different because we had incorrectly documented that it does support overlapping (where we are explicitly documenting that it doesn't).
https://github.com/dotnet/corefx/issues/34799

Alternatively, we could consider adding a guideline that all "OperationStatus" style APIs that take input/output spans should either document whether overlapping is supported or not (since they are generally for advance scenarios to begin with and perf sensitive), OR guard against overlaps explicitly in all cases.

cc @jkotas, @KrzysztofCwalina - what are your thoughts on that? Is documentation sufficient?

jkotas replied:
>  what are your thoughts on that? Is documentation sufficient?

I think so. The existing `EncodeUtf8` method on this type is on this plan already.
`

ahsonkhan commented on src/System.Text.Encodings.Web/tests/JavaScriptStringEncoderTests.cs:
```diff
@@ -18,11 +19,29 @@ public partial class JavaScriptStringEncoderTests
+        }
+
+        [Fact]
+        public void TestSurrogateBufferTooSmall()
```
How about empty input/output spans?

**These cases:**
empty input, empty output
empty input, non-empty output
non-empty input, empty output

Also, maybe a test that verifies that original contents of the destination are replaced (and only the contents in the expected range, with contents outside of the range remain in-tact).

steveharter replied:
Thanks - will do.

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs:
```diff
@@ -30,38 +30,43 @@ internal static partial class JsonWriterHelper
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static bool NeedsEscaping(char value) => value > LastAsciiCharacter || AllowList[value] == 0;
+        // Cast to (byte) for performance; avoids array bounds check.
```
It doesn't look like this will help and is even increasing asm size (due to a couple extra moves).
[sharplab.io](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGuf6GkAHa4MDAJZCxAGTwYAgrjDjxLABbYo2MBhhQGAXgYAGBAHYAYgG4+3WwwAOUcQDdsupuSQMASjGwAJgDyQgA2AJ4Ayg7YQgA8wOG6AHwM8qGhEADu0uKihqlCMFkMiboA2mUwAHQAstgIAGrYoQCuMAwA1AzkALr2PBwDgzzGaCbjYxPTU7OT8zMLc9MA9CsMAKqdxjvG1dVbuxbDI4tnyxdLV+fja5vbxuR7Bw/kx9Snp+Tj39O//9cAT0FkCgXdDsZSM8IaR3p8RqCfkjgSjEaiQRiGOCHnRoTi4fCeGjiciSejydidig8VSCYTyWTGcipmjKcYAKw0jl0wks0n8hkCplY9YQpBcpA8+HCmVC5m3UUPMxcyw0E4jO7pXAQCRCMBtAIdDBqDpaIQAcxguAYADMoBB8PcdgAOYwMDA6iEWCy26COPQ26D4WJgI067DEFCbAAqFmdggghtt9sdYBNYAA1pJzaUIK0hAFNOFqurBpcbhXy1W0KXRtdq5X61da9wG22m42W43u+2bl3ewOO3N+0PRz3mx96ePp4ORzOx722RYud71QBfGyTuxbxzONweYheXMQUIMAByMBgAVwAFElNgHNmABRVBhuNowACUBTSGWyuVEcp33aXpDCMYxN0GewnFcdwOkPbxgAgE9z0va87zAB9swAfSKLInzTTQ3xadpvwMVJgI6VJZFERRlFUDQtB0PQGEIQhf0yHI8gwcoXySL9KNAgxwMgnhoL3ODPEQ5DTwvK9b3vR8LWw6BxHNSQWgIxjiI/MiKJIqiGBohQlBUdRNG0XR9DYjj/24oCDKEkSaDXIA)

![image](https://user-images.githubusercontent.com/6527137/62241393-ee393d00-b38d-11e9-8ce9-42d8a524fdb7.png)

steveharter replied:
Thanks. I'll revert that

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs:
```diff
@@ -30,38 +30,43 @@ internal static partial class JsonWriterHelper
 
-            int idx;
-            for (idx = 0; idx < value.Length; idx++)
+            int len = value.Length;
```
Seems unnecessary to do this. Revert?

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs:
```diff
@@ -211,304 +219,28 @@ private static void EscapeNextBytes(byte value, Span<byte> destination, ref int
-
-        // We can't use the type Rune since it is not available on netstandard2.0
-        // To avoid extensive ifdefs and for simplicity, just using an int to reprepsent the scalar value, instead.
-        public static SequenceValidity PeekFirstSequence(ReadOnlySpan<byte> data, out int numBytesConsumed, out int rune)
```
Nice. I am glad this could be removed :)

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs:
```diff
@@ -211,304 +219,28 @@ private static void EscapeNextBytes(byte value, Span<byte> destination, ref int
-            // - Multi-byte sequences which are improperly terminated (no continuation byte when one is
-            //   expected) are reported as invalid sequences up to and including the last seen continuation byte.
-
-            Debug.Assert(JsonHelpers.IsValidUnicodeScalar(ReplacementChar));
```
Can we get rid of some of the other helpers that are no longer needed, like this one `JsonHelpers.IsValidUnicodeScalar`?

steveharter replied:
I'll check. I believe IsValidUnicodeScalar is used by NetFx, but could ifdef that then

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs:
```diff
@@ -211,304 +219,28 @@ private static void EscapeNextBytes(byte value, Span<byte> destination, ref int
-            EscapeString(utf8Destination.Slice(0, bytesWritten), utf8ConvertedDestination, indexOfFirstByteToEscape: 0, encoder, out int convertedBytesWritten);
+            Debug.Assert(result != OperationStatus.DestinationTooSmall);
+            Debug.Assert(result != OperationStatus.NeedMoreData);
+            Debug.Assert(encoderBytesConsumed == value.Length);
```
Would this be true if `OperationStatus != Done` but rather InvalidData? Looks like we should add a test to cover that scenario.

steveharter replied:
This look like an invalid Assert. The same for Utf8. I'll see if I can add a test for Utf8; utf16 currently will not return InvalidData

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs:
```diff
@@ -621,13 +348,13 @@ private static int WriteHex(int value, Span<char> destination, int written)
         {
             Debug.Assert(value < 16);
-            return (byte)((value < 10) ? ('0' + value) : ('a' + (value - 10)));
+            return (byte)((value < 10) ? ('0' + value) : ('A' + (value - 10)));
```
What motivated this change? Consistency with `JavascriptEncoder`?

steveharter replied:
Yes. Not related to this issue

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs:
```diff
@@ -529,28 +260,29 @@ public static void EscapeString(ReadOnlySpan<char> value, Span<char> destination
                 // first occurrence of a non-Ascii character, then call into the default encoder.
-                while (consumed < value.Length)
+                int len = value.Length;
+                while (indexOfFirstByteToEscape < len)
```
Revert caching `value.Length` in a local? Here/elsewhere.

ahsonkhan commented on src/System.Text.Json/tests/Utf8JsonWriterTests.cs:
```diff
@@ -3630,20 +3630,30 @@ public void EscapeSurrogatePairs(bool formatted, bool skipValidation)
+                jsonUtf8.WriteString(propertyName, value);
+                jsonUtf8.WriteEndObject();
+                jsonUtf8.Flush();
+                string result = Encoding.UTF8.GetString(
```
Use `JsonTestHelper.AssertContents`?

Also, consider moving this outside the using block (and then we can remove the call to Flush).

steveharter replied:
Sure I'll change it. I'll also add a "todo" for `AssertContents` and `AssertContentsNotEqual` since they assume Json.NET format and thus will do an unnecessary normalization when the expected value is not Json.Net.

ahsonkhan commented on src/System.Text.Json/tests/Utf8JsonWriterTests.cs:
```diff
@@ -3657,10 +3667,11 @@ public void EscapeSurrogatePairsThrowsHighSurrogateMissing()
             "{" + ValidUtf8 + ":" + ValidUtf8 + "}")]
         public void UTF8ReplacementCharacters(bool skipValidation, string expected0, string expected1, string expected2, string expected3)
         {
+            // SkipValidation does not affect whether we write the replacement character or not (we always do, unless we add a new option to control).
```
I wonder if this is worth documenting, maybe as a remark somewhere relevant. Seems too much of an edge case to do so on the option itself, but maybe it's good enough. Otherwise, maybe we add this remark to each of the `WriteXx` overloads that take strings.

steveharter replied:
Hmm the S.T.E.W API doc for Encode does not currently document this behavior.

Have you heard any asks for this functionality in the future and\or complaints about the previous behavior (which would throw instead of using the replacement char (like [this API](https://github.com/dotnet/corefx/blob/c5932104a914e4ec95a14a867da8bd109781e4e2/src/Common/src/CoreLib/System/Text/Unicode/Utf8.cs#L42) has via `replaceInvalidSequences` parameter)?

ahsonkhan replied:
> Hmm the S.T.E.W API doc for Encode does not currently document this behavior.

The SkipValidation concept is specific to the JSON APIs, not S.T.E.W. It is conceivable that someone would think that skipvalidation applies to replacement character vs leaving things as is, but I don't have any evidence of this. We wouldn't write invalid utf-8 as is anyway (we would either throw or replace). But your point is well taken, especially since I doubt anyone would care. No action needed.

ahsonkhan commented on src/System.Text.Json/tests/Utf8JsonWriterTests.cs:
```diff
@@ -3657,10 +3667,11 @@ public void EscapeSurrogatePairsThrowsHighSurrogateMissing()
 
-            var validUtf8 = new byte[2] { 0xc3, 0xb1 }; // 0xF1
             var invalidUtf8 = new byte[2] { 0xc3, 0x28 };
+            var validUtf8 = new byte[2] { 0xc3, 0xb1 }; // 0xF1
```
This change seems unnecessary to me.

ahsonkhan commented on src/System.Text.Json/tests/Utf8JsonWriterTests.cs:
```diff
@@ -3630,20 +3630,30 @@ public void EscapeSurrogatePairs(bool formatted, bool skipValidation)
 
         [Fact]
-        public void EscapeSurrogatePairsThrowsHighSurrogateMissing()
+        public void EscapeSurrogatePairs()
```
nit: `HighSurrogateMissingGetsReplaced`

ahsonkhan commented on src/System.Text.Json/tests/Utf8JsonWriterTests.cs:
```diff
@@ -3694,54 +3705,58 @@ string WriteProperty(byte[] propertyName, byte[] value)
+            "{" + InvalidUtf16 + ":" + ValidUtf16 + "}",
+            "{" + ValidUtf16 + ":" + InvalidUtf16 + "}",
+            "{" + ValidUtf16 + ":" + ValidUtf16 + "}")]
+        public void UTF16ReplacementCharacters(bool skipValidation, string expected0, string expected1, string expected2, string expected3)
```
nit: This seems like an odd test name.
Maybe `InvalidUTF16SurrogatePairsGetReplaced`

ahsonkhan replied:
Also, consider keeping the formatted/skipvalidation permutations.

steveharter replied:
I did not keep `formatted` permutations -- there's another test for that (HighSurrogateMissingGetsReplaced)

ahsonkhan commented on src/System.Text.Json/tests/Utf8JsonWriterTests.cs:
```diff
@@ -3694,54 +3705,58 @@ string WriteProperty(byte[] propertyName, byte[] value)
+                jsonUtf8.WriteString(propertyName, value);
+                jsonUtf8.WriteEndObject();
+                jsonUtf8.Flush();
+                string result = Encoding.UTF8.GetString(
```
Consider re-writing this to use `JsonTestHelper.AssertContents`

ahsonkhan commented on src/System.Text.Json/tests/Utf8JsonWriterTests.cs:
```diff
@@ -3694,54 +3705,58 @@ string WriteProperty(byte[] propertyName, byte[] value)
+            "{" + InvalidUtf16 + ":" + InvalidUtf16 + "}",
+            "{" + InvalidUtf16 + ":" + ValidUtf16 + "}",
+            "{" + ValidUtf16 + ":" + InvalidUtf16 + "}",
+            "{" + ValidUtf16 + ":" + ValidUtf16 + "}")]
```
nit: Make these separate test cases rather than several expected strings in one test

ahsonkhan commented:
> Improves performance of Utf16 encoding (with custom encode) by ~25% by avoiding Utf16->Utf8->Utf16 conversions:

> Also by changing code in System.Text.Json, improves performance of Utf8 escaping when no escaping needs to occur (~15% faster), or occurs later in the string

Great. We should get this in for 3.0, imo.

cc @ericstj 

> FWIW the ASCII escaping in System.Text.Json is ~25% faster than System.Text.Encoding.Web.JavaScriptEncoder - thus the reason why there is some duplicate functionality between the two; S.T.J will only call S.T.E.W for non-ASCII or when using a non-default encoder.

What's the reason for that? Is there a way for us to bring some of this perf improvement into S.T.E.W itself?

ahsonkhan commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs:
```diff
@@ -719,12 +780,17 @@ private byte[] GetAsciiEncoding(byte value)
+
+        private static void ThrowArgumentException_MaxOutputCharsPerInputChar()
+        {
+            throw new ArgumentException("Argument encoder does not implement MaxOutputCharsPerInputChar correctly.");
```
I think we generally put such exception message strings in a Strings resource file and use `SR.` to access them.

steveharter replied:
The S.T.E.W project currently does not use string resources.

ahsonkhan replied:
https://github.com/dotnet/corefx/issues/40074

steveharter commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs:
```diff
@@ -22,11 +22,12 @@ namespace System.Text.Encodings.Web
+
+        // Keep a reference to Array.Empty<byte> as this is used as a singleton for comparisons
+        // and there is no guarantee that Array.Empty<byte>() will always be the same instance.
+        private static readonly byte[] s_noEscape = Array.Empty<byte>();
```
We could have before, we just needed to keep a global reference to it. If we don't keep a global reference to it, then it is possible every call to Array.Empty returns a different instance which would cause reference equality to work incorrectly.

steveharter commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs:
```diff
@@ -524,6 +557,34 @@ internal static OperationStatus EncodeUtf8Shim(TextEncoder encoder, ReadOnlySpan
+        /// <summary>
+        /// Encodes the supplied characters.
+        /// </summary>
+        /// <param name="source">A source buffer containing the characters text to encode.</param>
```
Yes this was a typo. Thanks. However the other utf16 APIs here use "characters", so I'll change it to `A source buffer containing the characters to encode`

ahsonkhan replied:
Sure, that works too.

steveharter commented:
> cc @ericstj
> FWIW the ASCII escaping in System.Text.Json is ~25% faster than System.Text.Encoding.Web.JavaScriptEncoder - thus the reason why there is some duplicate functionality between the two; S.T.J will only call S.T.E.W for non-ASCII or when using a non-default encoder.
>> What's the reason for that? Is there a way for us to bring some of this perf improvement into S.T.E.W itself?

The reason S.T.E.W is slower is because there is a class hierarchy (JavaScriptEncoder : TextEncoder) where both the default encoder and custom share the same (slower) code which is driven off of a combination of valid and invalid ranges which are flexible and can be specified by the caller -- and this code can't have the micro-optimizations that we have in S.T.J.

However we could create a new internal type (DefaultJavaScriptEncoder in S.T.E.W) and move code from S.T.J to S.T.E.W however using a custom encoder is still going to be 25% slower since it won't use that fast code.

Also note that S.T.E.W was already made 3x-4x faster from the original (for Utf8).

GrabYourPitchforks commented on src/System.Text.Encodings.Web/src/System/Text/Unicode/UnicodeHelpers.cs:
```diff
@@ -363,7 +299,7 @@ private static int GetScalarValueFromUtf16Slow(char first, char? second, out boo
                 Debug.Assert(char.IsLowSurrogate(first));
                 wasSurrogatePair = false;
-                wasUnmatchedSurrogate = true;
+                needMoreData = true; // Last character was high surrgate; we need more data.
```
This line is incorrect. `needMoreData` should be _false_ if we see a standalone low surrogate character.

steveharter replied:
Thanks. Please verify with the latest commit.

GrabYourPitchforks commented:
It'd be good to have a unit test validating the changes for the latest iteration https://github.com/dotnet/corefx/pull/39900/commits/390e6c77528db10aef5dce208dcac2196903ef8e, but otherwise LGTM. Thanks!

ahsonkhan commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs:
```diff
@@ -607,7 +606,10 @@ internal static OperationStatus EncodeUtf8Shim(TextEncoder encoder, ReadOnlySpan
                     }
 
-                    return EncodeIntoBuffer(destinationPtr, destination.Length, sourcePtr, source.Length, out charsConsumed, out charsWritten, firstCharacterToEncode, isFinalBlock);
+                    fixed (char* destinationPtr = destination)
```
What would be the benefit of deferring creating this fixed block? In general, is it best to keep your fixed blocks as small as possible and is there no value in trying to batch them together with other fixed block setup?

cc @jkotas

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs:
```diff
@@ -48,12 +48,12 @@ internal static partial class JsonWriterHelper
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        // Cast to (byte) for performance; avoids array bounds check.
-        private static bool NeedsEscaping(char value) => value > LastAsciiCharacter || AllowList[(byte)value] == 0;
+        private static bool NeedsEscaping(char value) => value > LastAsciiCharacter || AllowList[value] == 0;
```
nit: implement this in terms of NeedsEscapingNoBoundsCheck

```C#
private static bool NeedsEscaping(char value)
    => value > LastAsciiCharacter || NeedsEscapingNoBoundsCheck(value);
```

ahsonkhan commented on src/System.Text.Json/tests/JsonTestHelper.cs:
```diff
@@ -696,6 +697,7 @@ public static void AssertContentsNotEqual(string expectedValue, ArrayBufferWrite
                     );
 
             // Temporary hack until we can use the same escape algorithm on both sides and make sure we want uppercase hex.
+            // Todo: create new AssertContentsNotEqualAgainJsonNet to avoid calling NormalizeToJsonNetFormat when not necessary.
```
Please create tracking issues for TODOs like this. They are easier to track/manage then searching for "todo" in source. Also, increases visibility for others to pick up issues like this.
